packages/aggregator/src/analytics/aggregateAnalytics/adjustOptionsToAggregationList.js

```diff
@@ -0,0 +1,134 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+import { convertDateRangeToPeriodString } from '@tupaia/utils';
+import { EntityConnection } from '../../connections';
+import { AGGREGATION_TYPES } from '../../aggregationTypes';
+import { getDateRangeForOffsetPeriod } from './aggregations/offsetPeriod';
+import { getDateRangeForSumPreviousPerPeriod } from './aggregations/sumPreviousPerPeriod';
+
+const AGGREGATION_TYPE_TO_DATE_RANGE_GETTER = {
+  [AGGREGATION_TYPES.SUM_PREVIOUS_EACH_DAY]: getDateRangeForSumPreviousPerPeriod,
+  [AGGREGATION_TYPES.OFFSET_PERIOD]: getDateRangeForOffsetPeriod,
+};
+
+/**
+ * Currently not all `startDate`, `endDate`, `period` options are provided in each client request.
+ * This function preserves the existing functionality by retaining the original period if
+ * there was no date range adjustment
+ */
+const getAdjustedPeriod = (originalOptions, adjustedOptions) => {
+  const areDateRangesEqual =
+    originalOptions?.startDate === adjustedOptions?.startDate &&
+    originalOptions?.endDate === adjustedOptions?.endDate;
+
+  return areDateRangesEqual
+    ? originalOptions.period
+    : convertDateRangeToPeriodString(adjustedOptions.startDate, adjustedOptions.endDate);
+};
+
+const getAdjustedDateRange = (dateRange, aggregationList) =>
+  aggregationList.reduce((currentRange, aggregation) => {
+    const getDateRange = AGGREGATION_TYPE_TO_DATE_RANGE_GETTER[aggregation.type];
+    return getDateRange ? getDateRange(currentRange, aggregation.config) : currentRange;
+  }, dateRange);
+
+const shouldFetchDataSourceEntities = ({ config: { dataSourceEntityType } = {} } = {}) =>
+  dataSourceEntityType;
+
+const shouldFetchRelations = ({
+  type,
+  config: { aggregationEntityType, dataSourceEntityType } = {},
+} = {}) => aggregationEntityType && dataSourceEntityType && !(type === AGGREGATION_TYPES.RAW);
```

Comment:
Similarly (and also optionally) here, you can consider optional chaining if it makes things clearer

Reply:
Yeah I agree, will do